### PR TITLE
Upload SBOM from jar file

### PR DESCRIPTION
**Pull request by eccles**
Problem:
SBOMS from jar files xmllint checks.

Solution:
Add optional -P flag to suppress validation and allow
either versiosn aor hashes.

Signed-off-by: Paul Hewlett <phewlett76@gmail.com>

### CODE REVIEW

File: scripts/sbom_scraper.sh [hunk 1]
```diff
@@ -53,18 +53,20 @@ then
     AUTHOR_EMAIL="$(git config user.email || true)"
     TOOL_NAME="$(git config --get remote.origin.url) $(git ls-files --full-name "$SCRIPTNAME")"
     TOOL_VERSION=$(git describe --tags)${GIT_STATUS:++}
+    SUPPLIER_NAME=github
```
Comment (j-hartley) [inline scripts/sbom_scraper.sh:56]: Let's remove this please, GitHub is not the SUPPLIER of the component.

Let's remove thus whole git machinery and require that AUTHOR_NAME,  AUTHOR_EMAIL and are provided as options.

And hard code TOOL_NAME to ${SCRIPTNAME}

TOOL_VERSION=$(git describe --tags)

File: scripts/sbom_scraper.sh [hunk 2]
```diff
@@ -53,18 +53,20 @@ then
...
+    SUPPLIER_URL=https://github.com
 else
     AUTHOR_NAME="$USER"
     AUTHOR_EMAIL=""
     TOOL_NAME="$SCRIPTNAME"
     TOOL_VERSION="unknown"
+    SUPPLIER_NAME=dockerhub
```
Comment (j-hartley) [inline scripts/sbom_scraper.sh:63]: move these to the conditional area below for docker image scraping

File: scripts/sbom_scraper.sh [hunk 3]
```diff
@@ -53,18 +53,20 @@ then
...
+    SUPPLIER_URL=https://hub.docker.com
 fi
 popd > /dev/null
 
 FORMAT=cyclonedx
 COMPONENT_AUTHOR_NAME="$AUTHOR_NAME"
-SUPPLIER_NAME=dockerhub
-SUPPLIER_URL=https://hub.docker.com
 TOOL_VENDOR="Jitsuin Inc"
 TOOL_HASH_ALG=SHA-256
```
Comment (j-hartley) [inline scripts/sbom_scraper.sh:71]: This is docker specific

File: scripts/sbom_scraper.sh [hunk 4]
```diff
@@ -160,16 +168,42 @@ then
...
-    log "Scrape ${FORMAT} SBOM from ${DOCKER_IMAGE} to ${OUTFILE} ..."
+    WORKDIR="${TEMPDIR}/jarfile"
+    mkdir "${WORKDIR}"
+    JARTYPE=$(echo "${DOCKER_IMAGE}" | cut -d':' -f1)
+    if [ "${JARTYPE}" = "https" ]
+    then
+        SUPPLIER_NAME=$(echo "${DOCKER_IMAGE}" | cut -d'/' -f4)
+        SUPPLIER_URL=$(echo "${DOCKER_IMAGE}" | cut -d'/' -f1-4)
+        (cd "${WORKDIR}" && curl -sSO "${DOCKER_IMAGE}")
+    else
```
Comment (j-hartley) [inline scripts/sbom_scraper.sh:186]: lets error out here

File: scripts/sbom_scraper.sh [hunk 5]
```diff
@@ -181,11 +215,17 @@ COMPONENT_NAME=${ORIG_COMPONENT_NAME%%:*}
 COMPONENT_VERSION=${ORIG_COMPONENT_NAME##*:}
```
Comment (j-hartley) [inline scripts/sbom_scraper.sh:215]: This is all docker image specific, move to conditional area above

File: scripts/sbom_scraper.sh [hunk 6]
```diff
@@ -181,11 +215,17 @@ COMPONENT_NAME=${ORIG_COMPONENT_NAME%%:*}
...
+    SHA256) COMPONENT_HASH_ALG="SHA-256"
+            ;;
+    *)      echo >&2 "Unknown hash algorithm $HASH_ALG"
+            COMPONENT_HASH_ALG=
+            ;;
 esac
 COMPONENT_HASH_CONTENT="${ORIG_COMPONENT_VERSION##*:}"
+if [ "$JARFILE" = "true" ]
+then
+    COMPONENT_VERSION=$(xq -r .bom.components.component.version  "${OUTPUT}")
```
Comment (j-hartley) [inline scripts/sbom_scraper.sh:227]: move into jar file conditional area above

File: scripts/sbom_scraper.sh [hunk 7]
```diff
@@ -220,13 +260,15 @@ echo "        content: $COMPONENT_HASH_CONTENT"
 [ -z "$TOOL_HASH_CONTENT" ] && echo >&2 "Unable to determine SBOM tool hash content" && exit 1
 [ -z "$AUTHOR_NAME" ] && echo >&2 "Unable to determine SBOM author name" && exit 1
 [ -z "$SUPPLIER_NAME" ] && echo >&2 "Unable to determine component supplier name" && exit 1
-[ -z "$SUPPLIER_URL" ] && echo >&2 "Unable to determine component supplier url" && exit 1
+#[ -z "$SUPPLIER_URL" ] && echo >&2 "Unable to determine component supplier url" && exit 1
```
Comment (j-hartley) [inline scripts/sbom_scraper.sh:263]: uncomment this

File: scripts/sbom_scraper.sh (hunk 4 above)
Comment (eccles) [reply scripts/sbom_scraper.sh:186]: moved logic to earlier in the script to error out before reaching here.

File: scripts/sbom_scraper.sh (hunk 3 above)
Comment (eccles) [reply scripts/sbom_scraper.sh]: fixed

File: scripts/sbom_scraper.sh (hunk 5 above)
Comment (eccles) [reply scripts/sbom_scraper.sh]: fixed

File: scripts/sbom_scraper.sh (hunk 6 above)
Comment (eccles) [reply scripts/sbom_scraper.sh]: fixed

File: scripts/sbom_scraper.sh (hunk 7 above)
Comment (eccles) [reply scripts/sbom_scraper.sh]: fixed

File: scripts/sbom_scraper.sh (hunk 2 above)
Comment (eccles) [reply scripts/sbom_scraper.sh]: done

File: scripts/sbom_scraper.sh [hunk 8]
```diff
@@ -41,39 +41,27 @@ log() {
 # ----------------------------------------------------------------------------
 # Option parsing
 # ----------------------------------------------------------------------------
+TOOL_NAME="https://github.com/jitsuin-inc/archib=vist-shell $SCRIPTNAME"
```
Comment (landintrees) [inline scripts/sbom_scraper.sh:44]: Typo: archib=vist-shell

File: scripts/sbom_scraper.sh [hunk 9]
```diff
@@ -82,7 +70,7 @@ usage() {
 
 Create a Cyclone DX 1.2 XML SBOM from a docker image and upload to RKVST SBOM Hub
 
-Usage: $SCRIPTNAME [-a AUTHOR_NAME] [-A AUTHOR_NAME] [-c CLIENT_SECRET_FILE] [-e AUTHOR_EMAIL] [-s] [-p] [-u URL] CLIENT_ID [docker-image:tag|sbom file]
+Usage: $SCRIPTNAME [-a AUTHOR_NAME] [-A AUTHOR_NAME] [-c CLIENT_SECRET_FILE] [-e AUTHOR_EMAIL] [-sp] [-u URL] CLIENT_ID [docker-image:tag|sbom file|jar URL]
```
Comment (landintrees) [inline scripts/sbom_scraper.sh:73]: Perhaps it would be better to explicitly use -A COMPONENT_AUTHOR here

File: scripts/sbom_scraper.sh (hunk 8 above)
Comment (eccles) [reply scripts/sbom_scraper.sh]: fixed

File: scripts/sbom_scraper.sh (hunk 9 above)
Comment (eccles) [reply scripts/sbom_scraper.sh]: fixed

**Review by landintrees — APPROVED** [review]
Look good, thanks

**Pull request merged**